nomos-da/network/core/src/sampling/behaviour.rs

```diff
@@ -0,0 +1,592 @@
+                        SamplingError::Deserialize {
+                            blob_id,
+                            subnetwork_id,
+                            peer_id,
+                            error,
+                        }
+                    });
+                match blob {
+                    Ok(blob) => Some(Poll::Ready(ToSwarm::GenerateEvent(
+                        SamplingEvent::SamplingSuccess {
+                            blob_id,
+                            subnetwork_id,
+                            blob: Box::new(blob),
+                        },
+                    ))),
+                    Err(error) => Some(Poll::Ready(ToSwarm::GenerateEvent(
+                        SamplingEvent::SamplingError { error },
+                    ))),
+                }
+            }
+            _ => {
+                error!("Invalid sampling response received, empty body");
+                None
+            }
+        }
+    }
+}
+
+impl<M: MembershipHandler<Id = PeerId, NetworkId = SubnetworkId> + 'static> NetworkBehaviour
+    for SamplingBehaviour<M>
+{
+    type ConnectionHandler = Either<
+        <libp2p_stream::Behaviour as NetworkBehaviour>::ConnectionHandler,
+        libp2p::swarm::dummy::ConnectionHandler,
+    >;
+    type ToSwarm = SamplingEvent;
+
+    fn handle_established_inbound_connection(
+        &mut self,
+        connection_id: ConnectionId,
+        peer: PeerId,
+        local_addr: &Multiaddr,
+        remote_addr: &Multiaddr,
+    ) -> Result<THandler<Self>, ConnectionDenied> {
+        if !self.membership.is_allowed(&peer) {
+            return Ok(Either::Right(libp2p::swarm::dummy::ConnectionHandler));
+        }
+        self.stream_behaviour
+            .handle_established_inbound_connection(connection_id, peer, local_addr, remote_addr)
+            .map(Either::Left)
+    }
+
+    fn handle_established_outbound_connection(
+        &mut self,
+        connection_id: ConnectionId,
+        peer: PeerId,
+        addr: &Multiaddr,
+        role_override: Endpoint,
+    ) -> Result<THandler<Self>, ConnectionDenied> {
+        if !self.membership.is_allowed(&peer) {
+            return Ok(Either::Right(libp2p::swarm::dummy::ConnectionHandler));
+        }
+        self.stream_behaviour
+            .handle_established_outbound_connection(connection_id, peer, addr, role_override)
+            .map(Either::Left)
+    }
+
+    fn on_swarm_event(&mut self, event: FromSwarm) {
+        self.stream_behaviour.on_swarm_event(event)
+    }
+
+    fn on_connection_handler_event(
+        &mut self,
+        peer_id: PeerId,
+        connection_id: ConnectionId,
+        event: THandlerOutEvent<Self>,
+    ) {
+        let Either::Left(event) = event else {
+            unreachable!()
+        };
+        self.stream_behaviour
+            .on_connection_handler_event(peer_id, connection_id, event)
+    }
+
+    fn poll(
```

Comment:
Would it be possible to reuse common parts from nomos behaviors to simplify `poll` and other network behavior methods?

Reply:
I thought about it. Some patterns are really similar, and probably we could make some common functions. But the small differences (types, reactions, etc) may be a burden later on to maintain IMO.